### DISCUSSION TITLE
Problems with long generated OpenAPI names and QRM

### DISCUSSION THREAD

**Discussion by Napam** [go-jet/jet#420] [Q&A] [answered]
### My scenario
In my application I write OpenAPI yaml first, then I generate models using oapi-codegen. For instance the return struct of the request `GET /listings/{id}/` gets generated to be called `GetListingsId200JSONResponse`. Generally speaking I would like to use said generated models as the destination for the QRM scanner. **This is itself is not a problem**, because I simply do something like:

```golang
columns := ProjectionList{}

for _, column := range t.Listings.AllColumns {
    columns = append(
        columns, 
        column.AS("GetListingsId200JSONResponse."+column.Name()),
    )
}

selectStatement := t.Listings.SELECT(
    columns,
).WHERE(
    t.Listings.ID.EQ(uuid),
)

var dest GetListingsId200JSONResponse

_ = selectStatement.Query(database.DB, &dest)
```
### The problem
And this works fine and looks nice and whatnot, I am very happy with it. BUT! Due to the long model names that gets generated, the expression `"GetListingsId200JSONResponse."+column.Name()`, can for example resolve to `GetListingsId200JSONResponse.this_here_is_a_kinda_long_column_name`. Now, this name will be longer than what postgres can handle, and the value that column name that postgres actually gives it is clipped, e.g. `GetListingsId200JSONResponse.this_here_is_a_kinda_long_column_na`, and that won't work for the QRM.

### Potential solutions I have thought of

Conceptually I want to "turn off" the requirement of specifying the struct name in the sql aliases, but after skimming throug the `go-jet` code I don't think I can do that as it is a very ingrained aspect of QRM, so:

#### Solution 1: Use shorter names (don't really like this idea)
A way to work around this is to have short column names, but I am not a fan of that if I have to trade off self descriptive column names. Also, some of the generated model names can get very long, which would force me to make really short column names, which I find unacceptable. Another way is to build postgres with another column size value (yes you have to build it), but I don't wanna do that.

#### Solution 2: Use the "type combination functionality"
I can do the following:
```golang
type Listing struct {
    GetListingsId200JSONResponse // generated by oapi-codegen
    this_here_is_a_kinda_long_column_name string
}
```
Then I can alias said column to "Listing.this_here_is_a_kinda_long_column_name", which will work. But I don't really want to do this as I will have to manually ensure that I do this for all columns that will be too long together with the struct name.

#### Solution 3: Create anon structs in runtime
A solution I can think of then is to make a "recursive anonymous struct creator" during runtime that basically takes in the dest, basically creates an equivalent struct but with no names, uses said struct for QRM, then I do a deepcopy of the anonymous struct to the actual struct that I want. Pseudo code:

```golang
columns := ProjectionList{}

for _, column := range t.Listings.AllColumns {
    columns = append(columns, column.AS("GetListingsId200JSONResponse."+column.Name()))
}

selectStatement := t.Listings.SELECT(
    columns,
).WHERE(
    t.Listings.ID.EQ(uuid),
)

var actualDest GetListingsId200JSONResponse
tempDest := CreateAnonStruct(&actualDest)

_ = selectStatement.Query(database.DB, &tempDest)

DeepCopy(&tempDest, &actualDest)
```

### My question
So my question is: is there any go-jet "native way" to deal with this, or do I just have to do the anonymization trick?

### What I know doesn't work
Type aliasing does not work, as the aliasing won't change the name when doing reflections, e.g.
`type Listing GetListingsId200JSONResponse` does not have the name `Listing` during reflection.

EDIT: I wrote "type aliasing does not work", but showed code for type declaration, basically
```golang
type Listing = GetListingsId200JSONResponse // I meant this
type Listing GetListingsId200JSONResponse // I wrote this, which is NOT aliasing, but a declaration (and this works! see answer by houten)
```

**Comment by Napam** [comment]
Another example of a long generated name that I have is:

`PostListingsListingIDImageContainersContainerIDUploadImageMultipartBody`

This name alone is longer than what postgres supports, so in theory I can't really use QRM for this model at all.

**Comment by houtn11** [comment, marked as answer]
> Type aliasing does not work, as the aliasing won't change the name when doing reflections, e.g.
type Listing GetListingsId200JSONResponse does not have the name Listing during reflection.

This should work, reflect.Type.Name() should return `Listing`.

**Reply by Napam** [reply]
Oh shoot, you're right, when I tried it the first time I did this:
```

type Listing = GetListingsId200JSONResponse
```
then it didn't work, and I just assumed that `type Listing GetListingsId200JSONResponse` would be the same.

But yeah, I just tried it now and it works. I should have explicitly tried this. Thank you very much of pointing me in the right direction (again)!
